File: 02. Algorithms/09. Dynamic Programming/009. Min Cost Climbing Stairs/code.cpp
class Solution {
public:
    
    int min(int a, int b) {
        return a<b?a:b ;
    }
    
    int solve(vector<int>cost, vector<int>&dp, int index) {
        if(index>=cost.size())
            return 0 ;
        
        if(dp[index]!=-1)
            return dp[index] ;
        
        return dp[index] = cost[index] + min(solve(cost,dp,index+1), solve(cost,dp,index+2)) ;
    }  
    
    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size() ;
        
        /*vector<int>dp(n+1,-1) ;
        return min(solve(cost,dp,0), solve(cost,dp,1)) ;*/
        
        int dp[n] ;
        
        dp[0] = cost[0] ;
        dp[1] = cost[1] ;
        
        for(int i=2;i<n;i++) {
            dp[i] = cost[i] + min(dp[i-1],dp[i-2]) ;
        }
        
        return min(dp[n-1],dp[n-2]) ;
    }
};

File: 02. Algorithms/09. Dynamic Programming/009. Min Cost Climbing Stairs/code.js
/**
 * Recursion
 * Time O(n) | Space O(n)
 * @param {number[]} cost
 * @return {number}
 */
var minCostClimbingStairs = function(cost) {
    
    const memo = {
        [cost.length - 1]: cost[cost.length - 1]
    };
    function dfs(index) {
        if(memo[index]) return memo[index];
        if(index >= cost.length) return 0;
        const choice1 = dfs(index+1);
        const choice2 = dfs(index+2);
        memo[index] = cost[index] + Math.min(choice1, choice2);
        return memo[index];
    }
    // console.log(memo);
    // console.log(memo);
    return Math.min(dfs(0), dfs(1));
    // return dfs(0);
};
